### DISCUSSION TITLE
Confused about best-practices for deployment with vercel + turborepo

### DISCUSSION THREAD

**Discussion by Tobjoern** [vercel/vercel#10185] [Help]
### Problem Description

Our enterprise is currently mirgrating our codebase to vercel, using a turborepo as our monorepo tool.
The development tooling works great, but we have some questions regarding the build + deployment process.

First of all, we use github actions, to build + test our application, in our pipeline itself.
After that, we intend to push them to their designated vercel projects, using the [vercel GH action](https://github.com/marketplace/actions/vercel-action),
in order to save resources, and not have unncessary builds, we intend to use the '[vercel deploy --prebuilt](https://vercel.com/guides/deploying-locally-built-nextjs)' command in the Github actions, for each project, if a change occured, in order to publish the latest changes.

Normally we validate the environment variables during the build process, but since we don't build the project on vercel itself, but just push the changes, we aren't sure how/when to do that?
Should we:
1. Load the vercel env in the GH-actions pipeline, and check it for validity
2. Create a custom check using Checkly in vercel, to 'test' the application once it is published?
3. Or accept the inefficiency, and build each application twice, once in GH actions, and once on vercel itself?

Furthermore, we are wondering how to handle the env inside a monorepo itself.
We currently have one single .env file in the root of the repository, should we create a project for the monorepo itself, and have all .env vars in a single project, or have them all stored for each project separately?
And if we decide to take the second approach, how would be pull them for nested projects?

Sorry for the long question, but I simply didn't find a good resources, which addresses all of these concerns and suggests best practices, cheers!

**Comment by g12i** [comment]
Thank you for this thread. 

I've been just setting similar pipeline myself. I'm using pnpm workspace + turborepo. My steps are something like:

```
checkout
setup node
pnpm install
pnpm run build -> turbo run build
vercel pull --yes --environment=${{ env.VERCEL_ENV }} --token=${{ }}
vercel build --token=${{ }}
vercel deploy --prebuilt --token=${{ }} >> $GITHUB_STEP_SUMMARY
```

I too need Vercel environment variables at some point, so after running `vercel pull` and before running my command  I do
```
set -o allexport; source .vercel/.env.${{ env.VERCEL_ENV }}.local; set +o allexport
```

Since running `turbo run build` builds Next App already, I added `"vercel-build": "echo \"Skipping build step\"",` script in my app's package.json so this step is avoided during vercel build [ref](https://github.com/vercel/vercel/blob/e5e757de34522cc4256296ffeed1b2e68664195a/packages/next/src/index.ts#L302)

Note that I have just one Vercel app in this monorepo so I kept _Root directory_ setting in Vercel and call vercel cli from the root directory. If I'd have more, I'd probably `cd` into each directory before calling vercel commands and leave _Root directory_ blank.
